Makefile: Add support for symbol versioning (ON by default)

topolarity opened the pull request:
This change enables symbol versioning by default and also adds `SYMBOL_VERSION_SUFFIX` as a Makefile option to allow appending a "salt" to the SONAME and all symbols' versions.

topolarity committed:
Makefile: Add support for symbol versioning (off by default)

This change adds `DEFAULT_SYMBOL_VERSION` as a Makefile option. If
specified, this will configure the linker to assign the specified
version string to all Julia-exported symbols.

The other half of the mechanism is `#define JL_SYMBOL_VERSION`, which
is used by the ccall machinery to look up symbols by their version + name.

topolarity commented on cli/Makefile:
```diff
@@ -104,14 +104,20 @@ julia-debug: $(build_bindir)/julia-debug$(EXE)
 libjulia-release: $(build_shlibdir)/libjulia.$(SHLIB_EXT)
 libjulia-debug: $(build_shlibdir)/libjulia-debug.$(SHLIB_EXT)
 
+ifeq ($(OS), Linux)
+VERSIONSCRIPT := -Wl,--version-script=$(BUILDDIR)/julia.generated.expmap
+else ifeq ($(OS), FreeBSD)
+VERSIONSCRIPT := -Wl,--version-script=$(BUILDDIR)/julia.generated.expmap
+endif
```
Nice trick, thanks!

topolarity commented on Make.inc:
```diff
@@ -180,6 +180,12 @@ JULIA_MAJOR_VERSION := $(shell echo $(JULIA_VERSION) | cut -d'-' -f 1 | cut -d'.
+# each must have a unique SONAME
+ifdef DEFAULT_SYMBOL_VERSION
+SOMAJOR := "$(SOMAJOR)-$(DEFAULT_SYMBOL_VERSION)"
+endif
```
Tested this morning - Julia-in-Julia is working as expected when used with PackageCompiler.jl

ChrisRackauckas commented:
Could we get a backport to v1.9.x here?

topolarity commented:
@staticfloat @gbaraldi  What do you think about turning this on by default?

If we go ahead and do that, I'm guessing we'll want a separate change that we won't backport to 1.9?

vtjnash reviewed (commented):
Can you simplify this PR to only have the version script change for the Makefile?

vtjnash commented on cli/Makefile:
```diff
@@ -144,8 +148,14 @@ $(build_bindir)/julia$(EXE): $(EXE_OBJS) $(build_shlibdir)/libjulia.$(SHLIB_EXT)
 $(build_bindir)/julia-debug$(EXE): $(EXE_DOBJS) $(build_shlibdir)/libjulia-debug.$(SHLIB_EXT) | $(build_bindir)
 	@$(call PRINT_LINK, $(CC) $(LOADER_CFLAGS) $(DEBUGFLAGS) $(EXE_DOBJS) -o $@ $(LOADER_LDFLAGS) $(RPATH) -ljulia-debug)
 
+$(BUILDDIR)/julia.generated.expmap:
+	@echo "$(DEFAULT_SYMBOL_VERSION)" > $@.tmp
+	@echo "{ global: *; };" >> $@.tmp
```
This appears to disregard the content of the existing version-script, which should be preserved
```suggestion
$(BUILDDIR)/julia.generated.expmap: $(SRCDIR)/julia.expmap
	@echo "$(DEFAULT_SYMBOL_VERSION)" > $@.tmp
	@cat "$<" >> $@.tmp
```

topolarity replied:
Ah, you'd like to re-use the `src/julia.expmap` as a template for the CLI loader version script?

Sounds good to me, I'll make that change

topolarity replied:
Actually, I misunderstood this request.

@staticfloat would you like me to add a `julia.expmap` to the `cli/` directory or to re-use the existing one in `src/`?

If the latter, what's the safe way to reach across directories in our Makefiles?

topolarity replied:
We attach the symbol versioning in both Makefiles - the difference is that `cli/` before this change didn't use any version script, so it was globally exporting all of its symbols by default.

It still does that with this change, but now it uses a linker script with `{ global: *; };`

topolarity commented:
> Can you simplify this PR to only have the version script change for the Makefile?

Sorry, I should have mentioned this in the OP, but f89b147aeb601b56f4a8684700afe8b68e29e278 is the only commit specific to this PR.

The others are carried from the dependency on #49010

topolarity commented:
Alright, I've added `cli/julia.expmap` and made the requested changes to #49010

This PR can now be reviewed independently 🙂

topolarity committed:
Makefile: use `sed` for version script generation

This changes the `.expmap` linker script generation to treat the source
file as a template and uses `sed` to replace strings in that template.

It also does a bit of re-naming, changing the template source to be
`julia.expmap.in` and the output to be `julia.expmap` instead of
`julia.generated.expmap`

topolarity commented on cli/julia.expmap:
```diff
@@ -0,0 +1 @@
+{ global: *; };
```
That's a great suggestion

> Sorry to keep suggesting changes, when overall this seems ready to go.

No worries - It's always a help to have your review :+1:

gbaraldi commented:
Bump!

vtjnash reviewed (commented):
LGTM. Can you make sure this is on-by-default, so that release builds have this available? I think you also need to specify `-fno-gnu-unique`, which disables a GNU extension to ELF as a feature to sometimes miscompile your C++ programs (https://bugs.llvm.org/show_bug.cgi?id=48221), and also check with @vchuravy that this won't break his fix in d8fa3c85600. IIUC, that symbol may need to be tagged with the LLVM symbol version, not the Julia symbol version info.

topolarity committed:
Add `-fno-gnu-unique` to JL_CFLAGS for gcc

topolarity commented:
> Can you make sure this is on-by-default, so that release builds have this available?

@vtjnash @staticfloat Sure thing. Does that mean we'd also like to change our SONAME format?

Presumably something like `libjulia.so.MAJOR.MINOR`?

topolarity committed:
Use LLVM's version string for `llvm::Any::TypeId`

topolarity committed:
Makefile: enable symbol versioning by default

This enables symbol versioning by default, with the version string
`JL_LIBJULIA_1.10` (the version suffix is the SOMAJOR in general).

As part of this change, the SOMAJOR has also been changed to be
MAJOR.MINOR instead of just MAJOR. This is required to allow multiple
minor releases of libjulia to live side-by-side in the same process.

`SYMBOL_VERSION_SUFFIX` is provided to allow appending an additional
unique "salt" to SOMAJOR, which can be helpful for creating template
binaries that will be "version-stamped" after distribution.

vtjnash commented on src/julia.expmap.in:
```diff
@@ -35,13 +35,16 @@
     *;
 };
+
+JL_LLVM_15.0 {
```
Can we get this from the Makefile instead? We have a preference for detection over declaration of versions, since it better supports upgrades and people who use system libraries. Something like this (probably could be made better and more reliable somehow though):
```
objdump -T libLLVM.so | grep _ZN4llvm3Any6TypeId | cut -f 2 | cut -d ' ' -f 3 | head -n 1
```
```
nm -D --with-symbol-versions libLLVM.so | grep _ZN4llvm3Any6TypeId | head -n 1 | sed -e 's/.*@//'
```

vtjnash replied:
`deps/llvm-ver.make` is the default version managed by deps, and is not the version of LLVM actually in use, so should not be used outside of that recipe

topolarity replied:
It seems like `LLVM_SHARED_LIB_NAME` or `LLVM_SHARED_LINK_FLAG` are both exported and used externally - Is that a bug?

topolarity replied:
I pushed a change with this approach, which works for me with `USE_BINARYBUILDER_LLVM=0` (and the usual build)

If there's a way to use a custom LLVM that can diverge from this, let me know and I'll take a different approach to fix it

topolarity committed:
Move LLVM_SHLIB_SYMBOL_VERSION to llvm-ver.make

Also, when `USE_SYSTEM_LLVM=1` compute `LLVM_SHLIB_SYMBOL_VERSION` using
`llvm-config`

vtjnash commented on cli/Makefile:
```diff
@@ -144,8 +148,13 @@ $(build_bindir)/julia$(EXE): $(EXE_OBJS) $(build_shlibdir)/libjulia.$(SHLIB_EXT)
 
+$(BUILDDIR)/julia.expmap: $(SRCDIR)/julia.expmap.in
+	sed <'$<' >'$@' -e 's/@JULIA_SHLIB_SYMBOL_VERSION@/JL_LIBJULIA_$(SOMAJOR)/' \
+		        -e 's/@LLVM_SHLIB_SYMBOL_VERSION@/$(LLVM_SHLIB_SYMBOL_VERSION)/'
```
Where is `deps/llvm-ver.make` coming from here? Those version numbers are the wrong values to be used outside of deps, and must instead be computed from the llvm-config-host library.

topolarity replied:
Ah I didn't know about `USE_SYSTEM_LLVM=1`. Yeah agreed, this is wrong when you're using system LLVM

It looks like we also use the wrong shlib name when updating `JL_PRIVATE_LIBS-0`:
https://github.com/JuliaLang/julia/blob/9d839f9794cbe4cc5eb654bcf8064f3f30452705/Makefile#L217-L219
